design/proposal-crossplane-v2.md

```diff
@@ -0,0 +1,948 @@
+only the composed resources in their own namespaces. This is more granular than
+today’s model. Today if the platform team wants to allow platform users to see
+their own RDSInstaces they must grant them access to read _all_ RDSInstances at
+the cluster scope - even those that belong to other teams.
+
+It would be equally valid for the platform team to decide that platform users
+have full RBAC access to Deployments, Services, and RDSInstances, making them
+the concern of the platform users. This granular model is more flexible to suit
+the needs of different organizations.
+
+Removing claims would have no effect on what a platform user can or can’t
+provision with Crossplane.
+
+A platform user with access to create a claim in a namespace is indirectly able
+to create the corresponding kind of cluster scoped XR. The claim pattern doesn’t
+impose any limitations on the shape or number of cluster scoped XRs the user can
+indirectly create. Therefore if a platform user needs to create cluster scoped
+XRs, giving them RBAC access to create them directly is functionally equivalent
+to giving them RBAC access to provision them indirectly via a claim.
+
+Removing support for claims would remove support for the late binding pattern.
+Late binding allows the platform team to pre-provision a cluster scoped XR,
+which is later claimed by a platform user. The claim doesn’t create a new XR
+on-demand - instead it binds to an existing one. In practice this feature is
+little known and seldom used.
+
+In order to remove support for claims (only) in the new v2 APIs, I propose that
+the new `spec.scope` CompositeResourceDefinition (XRD) field proposed in
+[Namespace Composite Resources](#namespace-composite-resources) would actually
+have three possible values:
+
+* Namespaced - Namespaced XR
+* Cluster - Cluster scoped XR without claim support
+* LegacyCluster - Cluster scoped XR with claim support
+
+The `spec.scope` field would default to LegacyCluster when using the v1 API.
+This ensures that behavior doesn’t change for existing Crossplane v1.x users
+when they upgrade to Crossplane 2.0.
+
+The `spec.scope` field would default to Namespaced when using the v2 API. The v2
+API would only allow the `spec.scope` field to be set to Namespaced or Cluster
+(not LegacyCluster) at creation time. The `spec.scope` field would be immutable
+once set.
+
+LegacyCluster must be a valid value in the v2 API in order to support
+Kubernetes’s round-trip constraint, which requires that it be possible to
+convert an API resource from v1 to v2 and back to v1 without data loss. This
+means LegacyCluster may appear when reading an XRD using the v2 API, but only if
+that XRD was created using the v1 API.
+
+The `spec.claimNames` and `spec.compositeDeletionPolicy` fields (which only
+apply to claims) must also appear in the v2 XRD API to satisfy the round-trip
+constraint. I propose both of these fields only be allowed to be set when
+`spec.scope` is `LegacyCluster`.
+
```

Comment:
How does this impact delete processing and forcing the use of foreground cascading deletion?  Removing the claim concept means that there is no longer a separation between the delete action taken by the claim owner and the deletion of the composite resource tree.  If the user doesn't explicitly specify `--cascade=foreground` when they delete the new claim/composite kubernetes will default to `background` and the deletion path will be uncontrolled except for `Usages`.  The `Usage` resource was never intended to take the place of standard kubernetes `foregound` deletion processing.

Reply:
I think we covered this in https://github.com/crossplane/crossplane/pull/6255/files#r1938188224. Let me know if I'm wrong.